File: labs/daemon/modules/relay/mesh_peer_matrix.go
package relay

import (
	"errors"
	"net"
	"sync"
	"time"
)

type PeerLinkMetric struct {
	DirectRttMs float64   `json:"direct_rtt_ms"`
	PacketLoss  float64   `json:"packet_loss"`
	Hops        uint32    `json:"hops"`
	Cost        float64   `json:"cost"`
	LastSeen    time.Time `json:"last_seen"`
}

func CalculateLinkCost(rtt float64, loss float64, hops uint32) float64 {
	// Cost = RTT * 0.7 + Loss * 350.0 + Hops * 15.0
	return (rtt * 0.7) + (loss * 350.0) + (float64(hops) * 15.0)
}

type MeshPeerTable struct {
	mu    sync.RWMutex
	peers map[string]map[string]*PeerLinkMetric
}

func NewMeshPeerTable() *MeshPeerTable {
	return &MeshPeerTable{
		peers: make(map[string]map[string]*PeerLinkMetric),
	}
}

func (m *MeshPeerTable) RecordLink(source, dest string, rtt float64, loss float64, hops uint32) {
	m.mu.Lock()
	defer m.mu.Unlock()

	if _, exists := m.peers[source]; !exists {
		m.peers[source] = make(map[string]*PeerLinkMetric)
	}
	cost := CalculateLinkCost(rtt, loss, hops)
	m.peers[source][dest] = &PeerLinkMetric{
		DirectRttMs: rtt,
		PacketLoss:  loss,
		Hops:        hops,
		Cost:        cost,
		LastSeen:    time.Now(),
	}
}

func (m *MeshPeerTable) FindBestRoute(source, dest string) (string, float64, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	adj, exists := m.peers[source]
	if !exists {
		return "", 0, errors.New("source peer unknown in mesh table")
	}

	bestNextHop := ""
	minCost := 1e9

	// Direct link
	if direct, ok := adj[dest]; ok {
		bestNextHop = dest
		minCost = direct.Cost
	}

	// 1-hop relay evaluation
	for intermediate, link1 := range adj {
		if intermediate == dest {
			continue
		}
		if interMap, ok := m.peers[intermediate]; ok {
			if link2, ok2 := interMap[dest]; ok2 {
				totalCost := link1.Cost + link2.Cost
				if totalCost < minCost {
					minCost = totalCost
					bestNextHop = intermediate
				}
			}
		}
	}

	if bestNextHop == "" {
		return "", 0, errors.New("no viable route found in mesh")
	}
	return bestNextHop, minCost, nil
}

type MeshNodeInfo struct {
	NodeID   string
	VirtualIP net.IP
}

File: labs/control-ui/api-shadow/mesh_peer_matrix.ts
export interface MeshLinkMetric {
  directRttMs: number;
  packetLoss: number;
  hops: number;
  cost: number;
}

export class MeshPeerMatrix {
  private readonly peers = new Map<string, Map<string, MeshLinkMetric>>();

  public recordLink(source: string, dest: string, rtt: number, loss: number, hops: number): void {
    const cost = rtt * 0.7 + loss * 350.0 + hops * 15.0;
    let adj = this.peers.get(source);
    if (!adj) {
      adj = new Map();
      this.peers.set(source, adj);
    }
    adj.set(dest, { directRttMs: rtt, packetLoss: loss, hops, cost });
  }

  public findBestRoute(source: string, dest: string): { nextHop: string; cost: number } | null {
    const adj = this.peers.get(source);
    if (!adj) return null;

    let bestNextHop: string | null = null;
    let minCost = Infinity;

    const direct = adj.get(dest);
    if (direct) {
      bestNextHop = dest;
      minCost = direct.cost;
    }

    for (const [intermediate, link1] of adj.entries()) {
      if (intermediate === dest) continue;
      const interMap = this.peers.get(intermediate);
      const link2 = interMap?.get(dest);
      if (link2) {
        const total = link1.cost + link2.cost;
        if (total < minCost) {
          minCost = total;
          bestNextHop = intermediate;
        }
      }
    }

    return bestNextHop ? { nextHop: bestNextHop, cost: minCost } : null;
  }
}
